commit 0934d640456bb168a8888ae388643c5160afe501	0934d640456bb168a8888ae388643c5160afe501
Author: drh <drh@noemail.net>
Date:   Mon May 25 15:19:52 2020 +0000

    Defensive code that tries to prevent a recurrence of problems like the
    one described in ticket [7a5279a25c57adf1]
    
    FossilOrigin-Name: 572105de1d44bca4f18c99d373458889163611384eebbc9659474874ee1701f4

diff --git a/manifest b/manifest
index f699358d3..a59fff912 100644
--- a/manifest
+++ b/manifest
@@ -1,5 +1,5 @@
-C Fix\sminor\scompile\sissues\swith\soptional\sextensions.
-D 2020-05-25T12:49:58.975
+C Defensive\scode\sthat\stries\sto\sprevent\sa\srecurrence\sof\sproblems\slike\sthe\none\sdescribed\sin\sticket\s[7a5279a25c57adf1]
+D 2020-05-25T15:19:52.511
 F .fossil-settings/empty-dirs dbb81e8fc0401ac46a1491ab34a7f2c7c0452f2f06b54ebb845d024ca8283ef1
 F .fossil-settings/ignore-glob 35175cdfcf539b2318cb04a9901442804be81cd677d8b889fcc9149c21f239ea
 F LICENSE.md df5091916dbb40e6e9686186587125e1b2ff51f022cc334e886c19a0e9982724
@@ -485,7 +485,7 @@ F src/date.c b29b349d277e3d579dcc295b24c0a2caed83fd8f090a9f7cbe6070c0fd662384
 F src/dbpage.c 8a01e865bf8bc6d7b1844b4314443a6436c07c3efe1d488ed89e81719047833a
 F src/dbstat.c 793deaf88a0904f88285d93d6713c636d55ede0ffd9f08d10f4ea825531d367f
 F src/delete.c 88047c8e59878c920fce14582bc1dde4d81157d1ca5ffdf36c2907e6d41996c4
-F src/expr.c 30bdb15abfa93d32862160795c3cb25d87489186ac0e34020eb2894767c80b42
+F src/expr.c b46669d9fc9e0361dba6cc289901a013789e0b1dc629c4c1bc88ec9403633b38
 F src/fault.c 460f3e55994363812d9d60844b2a6de88826e007
 F src/fkey.c 4b575423b0a5d4898b1a7868ce985cf1a8ad91c741c9abbb108ff02536d20f41
 F src/func.c 2333eb4277f55a5efdc12ef754e7d7ec9105d257b2fd00301d23ce1e8fa67dc0
@@ -1866,7 +1866,7 @@ F vsixtest/vsixtest.tcl 6a9a6ab600c25a91a7acc6293828957a386a8a93
 F vsixtest/vsixtest.vcxproj.data 2ed517e100c66dc455b492e1a33350c1b20fbcdc
 F vsixtest/vsixtest.vcxproj.filters 37e51ffedcdb064aad6ff33b6148725226cd608e
 F vsixtest/vsixtest_TemporaryKey.pfx e5b1b036facdb453873e7084e1cae9102ccc67a0
-P a77ceaf6ba934b1d73c90b8980191a65d02ad6ce1e11e6baf573b3a132685545
-R 4196412a957656021e354dac48b171f2
+P 3261ee9ec26bbeaa64190c12556a1d8146270b6660c70004830bd38c1b5b63f4
+R e34a1f962ac5984191c7a26f2bb9dcc0
 U drh
-Z 047ce2029fbbe737be54b5d1a4d5fdc1
+Z b86a7669719466f4efe7df019d717545
diff --git a/manifest.uuid b/manifest.uuid
index a5fa9fad9..2e0598a04 100644
--- a/manifest.uuid
+++ b/manifest.uuid
@@ -1 +1 @@
-3261ee9ec26bbeaa64190c12556a1d8146270b6660c70004830bd38c1b5b63f4
\ No newline at end of file
+572105de1d44bca4f18c99d373458889163611384eebbc9659474874ee1701f4
\ No newline at end of file
diff --git a/src/expr.c b/src/expr.c
index 83dd8b1ab..c5b678387 100644
--- a/src/expr.c
+++ b/src/expr.c
@@ -3811,7 +3811,10 @@ expr_code_doover:
   switch( op ){
     case TK_AGG_COLUMN: {
       AggInfo *pAggInfo = pExpr->pAggInfo;
-      struct AggInfo_col *pCol = &pAggInfo->aCol[pExpr->iAgg];
+      struct AggInfo_col *pCol;
+      assert( pAggInfo!=0 );
+      assert( pExpr->iAgg>=0 && pExpr->iAgg<pAggInfo->nColumn );
+      pCol = &pAggInfo->aCol[pExpr->iAgg];
       if( !pAggInfo->directMode ){
         assert( pCol->iMem>0 );
         return pCol->iMem;
@@ -4111,7 +4114,10 @@ expr_code_doover:
     }
     case TK_AGG_FUNCTION: {
       AggInfo *pInfo = pExpr->pAggInfo;
-      if( pInfo==0 ){
+      if( pInfo==0
+       || NEVER(pExpr->iAgg<0)
+       || NEVER(pExpr->iAgg>=pInfo->nFunc)
+      ){
         assert( !ExprHasProperty(pExpr, EP_IntValue) );
         sqlite3ErrorMsg(pParse, "misuse of aggregate: %s()", pExpr->u.zToken);
       }else{
